guppylang/prelude/_internal.py

```diff
@@ -15,93 +15,120 @@
 from guppylang.definition.ty import TypeDef
 from guppylang.definition.value import CallableDef
 from guppylang.error import GuppyError, GuppyTypeError
-from guppylang.hugr import ops, tys, val
-from guppylang.hugr.hugr import OutPortV
+from guppylang.hugr_builder.hugr import UNDEFINED, OutPortV
 from guppylang.nodes import GlobalCall
-from guppylang.tys.builtin import bool_type
+from guppylang.tys.builtin import bool_type, list_type
 from guppylang.tys.subst import Subst
 from guppylang.tys.ty import FunctionType, OpaqueType, Type, unify
 
 INT_WIDTH = 6  # 2^6 = 64 bit
 
 
-hugr_int_type = tys.Opaque(
-    extension="arithmetic.int.types",
-    id="int",
-    args=[tys.BoundedNatArg(n=INT_WIDTH)],
-    bound=tys.TypeBound.Eq,
+hugr_int_type = tys.Type(
+    tys.Opaque(
+        extension="arithmetic.int.types",
+        id="int",
+        args=[tys.TypeArg(tys.BoundedNatArg(n=INT_WIDTH))],
+        bound=tys.TypeBound.Eq,
+    )
 )
 
 
-hugr_float_type = tys.Opaque(
-    extension="arithmetic.float.types",
-    id="float64",
-    args=[],
-    bound=tys.TypeBound.Copyable,
+hugr_float_type = tys.Type(
+    tys.Opaque(
+        extension="arithmetic.float.types",
+        id="float64",
+        args=[],
+        bound=tys.TypeBound.Copyable,
+    )
 )
 
 
-class ConstIntS(BaseModel):
-    """Hugr representation of signed integers in the arithmetic extension."""
+class ConstInt(BaseModel):
+    """Hugr representation of signed and unsigned integers in the arithmetic extension.
+
+    Hugr always uses a u64 for the value. The interpretation is:
+      - as an unsigned integer, (value mod `2^N`);
+      - as a signed integer, (value mod `2^(N-1) - 2^(N-1)*a`)
+    where `N = 2^log_width` and `a` is the (N-1)th bit of `x` (counting from 0 = least
```

Comment:
Quite detailed! I was going to suggest that it'd be easier just to say `'a' is the most significant bit` but I realize that there are bits beyond the N. It *might* still be easier just to start with "take the N least significant bits of the value. If the value is unsigned...."

Or even, "the interpretation is the twos-complement of the N least significant bits".

I mean, this is correct, so ok ;), but also quite hard to follow.

Reply:
This is copied straight from Hugr repo:

https://github.com/CQCL/hugr/blob/1c422abd7d3eb0538635fd8a307de80944f93bcf/hugr/src/std_extensions/arithmetic/int_types.rs#L78-L88